server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceServlet.java

```diff
@@ -0,0 +1,163 @@
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.io.IOUtils;
+import org.apache.tomcat.util.http.fileupload.FileItem;
+import org.apache.tomcat.util.http.fileupload.disk.DiskFileItemFactory;
+import org.apache.tomcat.util.http.fileupload.servlet.ServletFileUpload;
+import org.sonar.api.server.ws.RailsHandler;
+import org.sonar.server.platform.Platform;
+
+public class WebServiceServlet extends HttpServlet {
+
+  private final Platform platform;
+  private WebServiceEngine webServiceEngine;
+  private Map<String, WsUrl> actions = new HashMap<>();
+
+  public WebServiceServlet() {
+    this.platform = Platform.getInstance();
+  }
+
+  @VisibleForTesting
+  WebServiceServlet(Platform platform) {
+    this.platform = platform;
+  }
+
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+    execute(req, resp);
+  }
+
+  @Override
+  protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+    execute(req, resp);
+  }
+
+  private void execute(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+    initUrls();
+    ServletRequest wsRequest = new ServletRequest(request, getExtraParameters(request));
+    ServletResponse wsResponse = new ServletResponse();
+
+    String path = request.getRequestURI().replaceFirst(request.getContextPath(), "");
+    String[] pathWithExtension = getPathWithExtension(path);
+    WsUrl url = actions.get(pathWithExtension[0]);
+    if (url != null) {
+      webServiceEngine.execute(wsRequest, wsResponse, url.getController(), url.getAction(), pathWithExtension[1]);
+
+      for (String header : wsResponse.getHeaderNames()) {
+        response.setHeader(header, wsResponse.getHeader(header));
+      }
+      response.setStatus(wsResponse.stream().httpStatus());
+      response.setContentType(wsResponse.stream().mediaType());
+
+      OutputStream responseOutput = response.getOutputStream();
+      ByteArrayOutputStream wsOutputStream = (ByteArrayOutputStream) wsResponse.stream().output();
+      IOUtils.write(wsOutputStream.toByteArray(), responseOutput);
+    }
+  }
+
+  private Map<String, Object> getExtraParameters(HttpServletRequest request){
+    Map<String, Object> maps = new HashMap<>();
+    if (ServletFileUpload.isMultipartContent(request)) {
+      // Create a factory for disk-based file items
+      DiskFileItemFactory factory = new DiskFileItemFactory();
+
+      // Configure a repository (to ensure a secure temp location is used)
+      ServletContext servletContext = this.getServletConfig().getServletContext();
+      // TODO use SonarQube temp dir
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)